gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/ConfigurableDispatch.java

```diff
@@ -41,13 +47,26 @@
 public class ConfigurableDispatch extends DefaultDispatch {
   private Set<String> requestExcludeHeaders = super.getOutboundRequestExcludeHeaders();
   private Set<String> responseExcludeHeaders = super.getOutboundResponseExcludeHeaders();
+  private Map<String, String> requestAppendHeaders = new HashMap<>();
   private Set<String> responseExcludeSetCookieHeaderDirectives = super.getOutboundResponseExcludedSetCookieHeaderDirectives();
   private Boolean removeUrlEncoding = false;
 
   private Set<String> convertCommaDelimitedHeadersToSet(String headers) {
     return headers == null ?  Collections.emptySet(): new HashSet<>(Arrays.asList(headers.split("\\s*,\\s*")));
   }
 
+  private Map<String, String> convertJSONHeadersToMap(String headers) {
```

Comment:
Putting JSON in XML looks wrong, we should follow the same pattern here as we follow elsewhere to extract parameters from service.xml file. I think NiFi dispatch does something like that with twoWaySSL.

Reply:
I had some doubts around this. I checked that in CustomDispatch, there is one xml param support:

@XmlAttribute(name = "use-two-way-ssl")
  public boolean getUseTwoWaySsl() {
    return useTwoWaySsl;
  }
and xml will look like this:
<requestAppendHeaders>
    <header>
      <name>a</name>
      <value>b</value>
    </header>
    <header>
      <name>c</name>
      <value>d</value>
    </header>
  </requestAppendHeaders>

Are you suggesting that we should have a new parameter requestAppendHeaders

or are you suggesting a layout like this:

<param>
  <requestAppendHeaders>
    <header>
      <name>a</name>
      <value>b</value>
    </header>
    <header>
      <name>c</name>
      <value>d</value>
    </header>
  </requestAppendHeaders>
</param>

Reply:
Analysed both the approaches which I posted above, first one looks like we are allowing a new special variable which does not come under param scope, that looked a little bad to me
Second one however looks functionally more correct but will require many changes in param structure in  CustomDispatch class, which currently allows only name, value pairs inside it

Not sure what should be done here

Reply:
The twoWaySSL is implemented as attribute :( 
I was thinking using <params> to keep the XML generic, something like 
` <param>
            <name>requestAppendHeaders</name>
            <value>a:b,c:d</value>
        </param>`

Now, that brings up an interesting point, what happens if the header contains `:` how do we handle it?

Reply:
Or we could have a attribute `requestAppendHeaders` on `param` and use name value.

Reply:
Sorry for the delay in response here, I have pushed a new commit. What I am trying to do is:

Support for a values inside DispatchParam. So either a param can be a simple param like:
<param>
<name>a</name>
<value>b</value>
</param>

Or it can be a complex param like this:

<param>
<name>a</name>
<values>
  <param>
    <name>a1</name>
    <value>b1</value>
  </param>
  <param>
    <name>a2</name>
    <value>b2</value>
  </param>
</values>
</param>

Reply:
I have not added test cases for these yet, please let me know if this sounds fine to you

Reply:
Apologies for the delay, the simple param example looks good to me, I cannot think of a case where we need complex param, do you have one in mind? AFA headers are concerned they are simple <key><value> so I do not think we need to complicate this, now depending on header the separator might vary (e.g. SET-COOKIE header) but that should not be a problem here. We could do a simple validation to make sure that the header values don't break the request. 

Test would be great to have we don't this to break accidentally due to any later change.
Thanks.

Reply:
yeah I also think the same, I felt that I am complicating the code but was not sure about your view. So, now can I implement something like this: <param> <name>requestAppendHeaders</name> <value>a:b,c:d</value> </param>

Reply:
I will add relevant checks and guardrails with logs

Reply:
Sounds good, just to clarify what I mean following are few examples
```
<param> 
     <name>requestAppendHeaders</name> 
     <value>Transfer-Encoding:gzip, chunked; Keep-Alive: timeout=5, max=1000</value> 
</param>
```
Here `;` is used as separator BUT `;` is used in SET-COOKIE header e.g.
`Set-Cookie: <cookie-name>=<cookie-value>; Domain=<domain-value>; Secure; HttpOnly`

The question here is should we be allowing to set headers like SET-COOKIE? if yes how do we handle `;` 
one way would be to use an escape `\;` 

Also, I think your patch is for inbound requests only right? i.e. requests coming in from client to Knox and knox setting headers on requests it dispatches, not the other way around. If it is both ways then we might have some issues. 
See, now this is starting to look complicated :)

Reply:
@moresandeep  I have done all the changes from my side, please check once. Added testcases also

Reply:
Thank you @mudit-97 I will build the patch and check it out, thanks for quick turnaround.

Reply:
@moresandeep  let me know if any more changes are required

Reply:
@moresandeep  did you get time to check this?, please let me know if further changes are required

Reply:
Sorry @mudit-97 got buried in other stuff, this is still on top of my todo list. Last week I tried to build a docker container for your patch (https://github.com/moresandeep/knox-dev-docker) but it failed, so i'll have to manually download the patch and check out.

Reply:
Thanks for the update

Reply:
@moresandeep , I tried testing this on my local, that was working fine. I tested one positive and one negative case, let me know if you face any issues